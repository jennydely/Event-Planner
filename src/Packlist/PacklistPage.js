import React from 'react'
import styled from 'styled-components/macro'
import { useParams, useHistory } from 'react-router-dom'
import usePacklists from './usePacklists'
import ListItem from '../common/ListItem'
import ListContainer from '../common/ListContainer'
import Checkbox from '../common/Checkbox'
import { comparePacklists } from '../services/comparePacklists'
import editIcon from '../img/editIcon.svg'
import backIcon from '../img/backIcon.svg'

export default function PackListPage() {
  const { packlistName } = useParams()
  const { packlists, updatePacklistCheckbox } = usePacklists()
  const chosenPacklist = comparePacklists(packlists, packlistName)
  const history = useHistory()
  function goBackButton() {
    history.goBack()
  }
  function handleEditPacklist(packlistId) {
    history.push('/editpacklist/' + packlistId)
  }
  return (
    <>
      <header>
        {packlistName ? <h1>PackList</h1> : <h1>No PackList</h1>}
        <button onClick={handleEditButtonClick}>
          <EditImg src={editIcon} alt="edit" />
        </button>
      </header>
      <main>
        {packlistName ? <PacklistButton>{packlistName}</PacklistButton> : ''}

        {chosenPacklist ? (
          <ListContainer>
            {chosenPacklist.packlist.sort().map((item) => (
              <ListItemStyled key={item.itemID} id={item.itemID}>
                <Checkbox
                  type="checkbox"
                  checked={item.checked}
                  onChange={() => handleCheckboxClick(item)}
                />
                <span>{item.item}</span>
                <div></div>
              </ListItemStyled>
            ))}
          </ListContainer>
        ) : (
          <NoPacklistText>
            There is no packlist added to this event
          </NoPacklistText>
        )}
      </main>
      <>
        <footer>
          <button type="button" onClick={goBackButton}>
            <img src={backIcon} alt="back" />
          </button>
        </footer>
      </>
    </>
  )
  function handleEditButtonClick() {
    const packlistId = chosenPacklist.id
    handleEditPacklist(packlistId)
  }

  function handleCheckboxClick(checkedItem) {
    const id = checkedItem.itemID
    const index = chosenPacklist.packlist.findIndex(
      (item) => item.itemID === id
    )
    const clickedPacklistItem = chosenPacklist.packlist[index]
    clickedPacklistItem.checked = !clickedPacklistItem.checked
    updatePacklistCheckbox(chosenPacklist)
  }
}

const EditImg = styled.img`
  width: 40px;
  height: auto;
  margin: 7px 4px;
`
const NoPacklistText = styled.p`
  text-align: center;
`
const ListItemStyled = styled(ListItem)`
  justify-content: flex-start;
`
const PacklistButton = styled.button`
  border: var(--border-darkgrey);
  margin: 4px 0 0 7px;
  padding-bottom: 0;
`
