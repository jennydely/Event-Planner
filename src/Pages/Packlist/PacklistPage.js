import React, { useContext, useEffect } from 'react'
import styled from 'styled-components/macro'
import { useParams, useHistory } from 'react-router-dom'
import editIcon from '../../img/editIcon.svg'
import FormHeader from '../components/FormHeader'
import ListItem from '../components/common/ListItem'
import ListContainer from '../components/common/ListContainer'
import Checkbox from '../components/common/Checkbox'
import { comparePacklists } from './services/comparePacklists'
import usePacklists from './usePacklists'
import fetchPrivatePacklists from '../../redux/fetchPrivatePacklists'
import fetchPublicPacklists from '../../redux/fetchPublicPacklists'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { UserContext } from '../../providers/UserProvider'

export function PacklistPage({ packlists, fetchPrivatePacklists, fetchPublicPacklists }) {
  const user = useContext(UserContext);
  const { packlistName } = useParams()
  const { updatePacklistCheckbox } = usePacklists()
  const chosenPacklist = comparePacklists(packlists, packlistName)
  const history = useHistory()

  function handleEditPacklist(packlistId) {
    history.push('/packlistform/' + packlistId)
  }

  useEffect(() => {
    fetchPublicPacklists()
    if (user) {

      fetchPrivatePacklists({ userToken: user.token })
    }
  }, [fetchPublicPacklists, fetchPrivatePacklists, user])


  return (
    <>
      <FormHeader
        headerText={packlistName ? 'Packlist' : 'No PackList'}
        headerButton={
          packlistName ? (
            <EditButton onClick={handleEditButtonClick}>
              <img src={editIcon} alt="edit" />
            </EditButton>
          ) : (
              ''
            )
        }
      />
      <main>
        {packlistName ? <PacklistName>{packlistName}</PacklistName> : ''}

        {chosenPacklist ? (
          <ListContainer>
            {chosenPacklist.packlist.sort().map((item) => (
              <ListItemStyled key={item.itemID} id={item.itemID}>
                <ItemContainer>
                  <Checkbox
                    type="checkbox"
                    checked={item.checked}
                    onChange={() => handleCheckboxClick(item)}
                  />
                  {item.item}
                </ItemContainer>
              </ListItemStyled>
            ))}
          </ListContainer>
        ) : (
            <NoPacklistText>
              There is no packlist added to this event
            </NoPacklistText>
          )}
      </main>
    </>
  )
  function handleEditButtonClick() {
    const packlistId = chosenPacklist.id
    handleEditPacklist(packlistId)
  }

  function handleCheckboxClick(checkedItem) {
    const id = checkedItem.itemID
    const index = chosenPacklist.packlist.findIndex(
      (item) => item.itemID === id
    )
    const clickedPacklistItem = chosenPacklist.packlist[index]
    clickedPacklistItem.checked = !clickedPacklistItem.checked
    updatePacklistCheckbox(chosenPacklist)
  }
}

const mapStateToProps = state => {
  const { publicPacklists } = state.packlists || {};
  return { packlists: publicPacklists };
};
const mapDispatchToProps = dispatch => bindActionCreators({
  fetchPrivatePacklists: fetchPrivatePacklists,
  fetchPublicPacklists: fetchPublicPacklists
}, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(PacklistPage)

const NoPacklistText = styled.p`
  text-align: center;
`
const ListItemStyled = styled(ListItem)`
  justify-content: flex-start;
`
const PacklistName = styled.h2`
  text-align: left;
  width: fit-content;
  border: var(--borderUL);
  border-radius: 7px;
  margin: 4px 0 0.13px 7px;
  padding: 0 4px;
`
const ItemContainer = styled.label`
  display: flex;
  align-items: center;
`

const EditButton = styled.button`
  margin: 0 7px;
`
