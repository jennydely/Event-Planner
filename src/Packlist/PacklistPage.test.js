import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { render } from '@testing-library/react'
import renderer from 'react-test-renderer'
import PacklistPage from './PacklistPage'
import 'jest-styled-components'

describe('PackList', () => {
    const packlistName = "festival";
    const packlists = {
        "id": 500,
        "name": "festival",
        "packlist": 
           ["Wasser",
            "Plastikzelt",
            "Campingkocher",
            "Essen",
            "Pfanne & Öl",
            "Cola",
            "Festivalbox I",
            "Festivalbox II",
            "Ticket",
            "Waschschale",
            "Teppich",
            "Isomatte",
            "Luftmatratze",
            "Bettdecke & Kissen",
            "Knicklichter",
            "Alu-Herringe",
            "Powerbank & Ladekabel",
            "Kopftuch",
            "Portmornai mit Bargeld",
            "Kulturtasche",
            "Handtücher",
            "Kleidung",
            "Cider",
            "Metalgürtel",
            "ausgedruckter Timetable"
        ]}

    it('displays the packlist festival', () => {
        const { getByText } = render(
        <MemoryRouter>
        <PacklistPage packlists={packlists} packlistName={packlistName} />
        </MemoryRouter>
        )
        setTimeout(() => {
            expect(getByText(packlistName)).toBeInTheDocument()
            expect(getByText('Wasser')).toBeInTheDocument()
        }, 100)
    })
it('renders correctly', () => {
    const tree = renderer.create(
        <MemoryRouter>
            <PacklistPage packlists={packlists} packlistName={packlistName} />
        </MemoryRouter>)
    expect(tree).toMatchSnapshot()
})
})