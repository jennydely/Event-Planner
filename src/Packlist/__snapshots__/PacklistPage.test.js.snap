// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PackList renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  width: 40px;
  height: auto;
  margin: 7px 4px;
}

.c1 {
  text-align: center;
}

<body>
    <div>
      <header>
        <h1>
          No PackList
        </h1>
        <button
          class="PacklistPage__EditButton-t2ih9w-1"
        >
          <img
            alt="edit"
            class="c0"
            src="editIcon.svg"
          />
        </button>
      </header>
      <main>
        
        <p
          class="c1"
        >
          There is no packlist added to this event
        </p>
      </main>
      <footer>
        <button
          type="button"
        >
          <img
            alt="back"
            src="backIcon.svg"
          />
        </button>
      </footer>
    </div>
  </body>,
  "container": <div>
    <header>
      <h1>
        No PackList
      </h1>
      <button
        class="PacklistPage__EditButton-t2ih9w-1"
      >
        <img
          alt="edit"
          class="PacklistPage__EditImg-t2ih9w-0 fORUOQ"
          src="editIcon.svg"
        />
      </button>
    </header>
    <main>
      
      <p
        class="PacklistPage__NoPacklistText-t2ih9w-2 bFfQZL"
      >
        There is no packlist added to this event
      </p>
    </main>
    <footer>
      <button
        type="button"
      >
        <img
          alt="back"
          src="backIcon.svg"
        />
      </button>
    </footer>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
